blog/_posts/2018-06-10-bsp.md

```diff
@@ -0,0 +1,214 @@
+are complementary by design.
+
+While LSP specifies endpoints for communication between an *editor acting as
+a client* and a language server, BSP specifies endpoints between a *language
+server acting as a client* and a build server.
+
+It turns out that if we want to support Scala across editors like Visual
+Studio Code, `vim` or Sublime via a Scala language server, we also need to
+integrate with build tools to support features like compiler diagnostics or
+code navigation. One of the main focus of BSP is that it can be implemented
+alongside LSP.
+
+## The Build Server Protocol v1.0.0
+
+Today I announce version `v1.0.0` of the Build Server Protocol. This version
+has been the result of a fruitful collaboration with [Justin
+Kaeser][jastice], the [JetBrains] developer responsible for sbt support in
+[IntelliJ IDEA].
+
+Justin and I have spent the past months implementing a prototype of the Build
+Server Protocol between [IntelliJ IDEA] (client) and [Bloop] (server). The
+goal of this prototype was two-fold. On the one hand, we wanted to experience
+first-hand the challenges of implementing BSP and learn new insights to
+improve the protocol. On the other hand, we wanted to provide Scala
+developers a faster "Import Project" that would work across different build
+tools *now*, even though build tools don't yet provide implementations of
+build servers. Thanks to Bloop, users of [sbt], [mill] or [Maven] ---build
+tools that Bloop currently supports--- can try this prototype out.
+
+[BSP v1.0.0](https://github.com/scalacenter/bsp/blob/master/docs/bsp.md)
+has been designed to support the most basic language integration in an
+IDE/editor. IDEs can find all the modules defined in a workspace, ask for
+their dependencies, compile, test, run, and get notifications on modified
+build targets, among a few other features.
+
+Our IntelliJ-Bloop implementation doesn't yet implement all these endpoints.
+Instead, it focuses on providing a faster integration than the traditional
+project import. You can now import your project, browse Scala code with full
+language support and get compiler diagnostics from the build tool in
+IntelliJ's editor.
+
+How does this work? First, you export your project from your build to Bloop.
+This process takes usually ~10s for a medium-sized project (highly
+recommended to use [coursier] in your build for upmost speed) and exists so
+that Bloop can understand different build tools. Then, the IDE imports it
+from Bloop via BSP. In total, the "Import Project" process takes around ~15
+seconds.
+
+You can see it on this demo importing [akka/akka](https://github.com/akka/akka/):
+
+![Demo of BSP import for akka/akka](/resources/img/bloop-install.gif)
+
+In my experience, importing a build via BSP is substantially faster than
+doing so via the traditional sbt or Maven project import where the process
+can take minutes. It also enables smarter communication between build tools
```

Comment:
I’m not sure this is fair. IntelliJ opens sbt project in instant time, and then all the project’s dependencies are downloaded in the background. You don’t have to wait for minutes to start working.

Reply:
I'm not sure we're talking about the same thing. Are you talking about the workflow after you've imported the project, you close IntelliJ and then you open it up again? Project's dependencies are not downloaded by IntelliJ but by sbt during the import process and the process to do so is slow even if the dependencies are already downloaded. For example, import zinc in IntelliJ (just tried it) takes 2 minutes! That's the price I need to pay every time I update a dep in my build so that the IDE knows about it.

Reply:
(sorry to pollute the PR with that discussion…)

No, I’m talking about importing an sbt for the first time in IntelliJ. I’ve just tried with zinc and while the project is indeed imported in the background (this has taken one minute) you can still work on it. That’s why I think it’s not fair to formulate things in a way that make people believe that you will cut several minutes of waiting.

Reply:
depends, the instant opening only sort-of works for simple projects with standard layout and provides only the editor

Reply:
> No, I’m talking about importing an sbt project for the first time in IntelliJ. I’ve just tried with zinc and while the project is imported in the background (this has indeed taken one minute)